File: tests/test_numpy_dtypes_user.cpp
/*
  tests/test_numpy_dtypes.cpp -- User defined NumPy dtypes

  Copyright (c) 2018 Eric Cousineau

  All rights reserved. Use of this source code is governed by a
  BSD-style license that can be found in the LICENSE file.
*/

// TODO(eric.cousineau): See what mixing non-dtypes look like for pybind...

#include <cstring>

#include "pybind11_tests.h"
#include "constructor_stats.h"
#include <pybind11/stl.h>
#include <pybind11/operators.h>
#include <pybind11/numpy_dtypes_user.h>

using std::string;
using std::unique_ptr;

namespace py = pybind11;

namespace {

// Trivial string class.
class CustomStr {
public:
    static constexpr int len = 100;
    CustomStr(const char* s) {
        snprintf(buffer.data(), len, "%s", s);
        dummy.reset(new int(1000));
    }
    template <typename Arg, typename... Args>
    CustomStr(const char* fmt, Arg arg, Args... args) {
        snprintf(buffer.data(), len, fmt, arg, args...);
        dummy.reset(new int(1000));
    }
    CustomStr(const CustomStr&) = default;
    CustomStr& operator=(const CustomStr&) = default;
    CustomStr(CustomStr&&) = default;
    CustomStr& operator=(CustomStr&&) = default;
    std::string str() const {
        return buffer.data();
    }
    bool operator==(const CustomStr& other) const {
        return str() == other.str();
    }
private:
    std::array<char, len> buffer;
    // Data member to ensure that we do not get segfaults when carrying around `shared_ptr`s.
    std::shared_ptr<int> dummy;
};

// Basic structure, meant to be an implicitly convertible value for `Custom`.
// Would have used a struct type, but the scalars are only tuples.
struct SimpleStruct {
    double value;

    SimpleStruct(double value_in) : value(value_in) {}
};

// Define operations to return fixed values.
double operator+(double, SimpleStruct) {
    return -1;
}
double operator+(SimpleStruct, SimpleStruct) {
    return 0;
}
double operator+(SimpleStruct, double) {
    return 1;
}

// Clones a unique_ptr using type's copy constructor.
template <typename T>
void clone(const unique_ptr<T>& src, unique_ptr<T>& dst) {
    if (!src)
        dst.reset();
    else
        dst.reset(new T(*src));
}

class Custom {
public:
    Custom() {
        print_created(this);
    }
    ~Custom() {
        track_destroyed(this);
        // TODO(eric.cousineau): Track down why this causes a segfault.
        //print_destroyed(this);
    }
    Custom(double value) : value_{value} {
        print_created(this, value);
    }
    Custom(double value, string str)
        : value_{value}, str_{unique_ptr<string>(new string(str))} {
        print_created(this, value, str);
    }
    Custom(const Custom& other) {
        value_ = other.value_;
        clone(other.str_, str_);
        print_copy_created(this, other.value_);
    }
    Custom(const SimpleStruct& other) {
        value_ = other.value;
        print_copy_created(this, other);
    }
    Custom& operator=(const Custom& other) {
        print_copy_assigned(this, other.value_);
        value_ = other.value_;
        clone(other.str_, str_);
        return *this;
    }

    double value() const { return value_; }

    operator SimpleStruct() const { return {value_}; }

    Custom operator+(const Custom& rhs) const {
        auto tmp = Custom(value_ + rhs.value_);
        return tmp;
    }
    Custom& operator+=(const Custom& rhs) {
        value_ += rhs.value_;
        return *this;
    }
    Custom operator+(double rhs) const {
        auto tmp = Custom(value_ + rhs);
        return tmp;
    }
    Custom& operator+=(double rhs) {
        value_ += rhs;
        return *this;
    }
    Custom operator*(const Custom& rhs) const { return value_ * rhs.value_; }
    Custom operator-(const Custom& rhs) const { return value_ - rhs.value_; }

    Custom operator-() const { return -value_; }

    Custom operator/(const Custom& rhs) const {
        return Custom(value_ / rhs.value_);
    }

    bool same_as(const Custom& rhs) const {
        return value_ == rhs.value_ && str() == rhs.str();
    }
    CustomStr operator==(const Custom& rhs) const {
        // Return non-boolean dtype.
        return CustomStr("%g == %g && '%s' == '%s'", value_, rhs.value_, str().c_str(), rhs.str().c_str());
    }
    bool operator<(const Custom& rhs) const {
        // Return boolean value.
        return value_ < rhs.value_;
    }

    std::string str() const {
        if (str_)
            return *str_;
        else
            return {};
    }

private:
    double value_{};
    // Use non-trivial data object, but something that is memcpy-movable.
    std::unique_ptr<string> str_;
};

Custom operator+(double a, const Custom& b) {
    return Custom(a) += b;
}

// Define `dtype=object` types, and ensure the operations correctly reflect these.
struct ObjectA {};
struct ObjectB {};

Custom operator+(const Custom& a, const ObjectA& b) {
    return Custom(1000);
}
Custom operator+(const Custom& a, const ObjectB& b) {
    return Custom(9999);
}

}  // namespace

#if defined(PYBIND11_CPP14)

PYBIND11_NUMPY_DTYPE_USER(CustomStr);
PYBIND11_NUMPY_DTYPE_USER(SimpleStruct);
PYBIND11_NUMPY_DTYPE_USER(Custom);

PYBIND11_NUMPY_OBJECT_DTYPE(ObjectA);
PYBIND11_NUMPY_OBJECT_DTYPE(ObjectB);

TEST_SUBMODULE(numpy_dtype_user, m) {
    ConstructorStats::type_fallback([](py::object cls) -> std::type_index* {
        return const_cast<std::type_index*>(py::detail::dtype_info::find_entry(cls));
    });

    try { py::module::import("numpy"); }
    catch (...) { return; }

    py::class_<ObjectA>(m, "ObjectA")
        .def(py::init());
    py::class_<ObjectB>(m, "ObjectB")
        .def(py::init());

    // Bare, minimal type.
    py::dtype_user<CustomStr>(m, "CustomStr")
        .def(py::init<const char*>())
        .def("__str__", &CustomStr::str)
        .def("__repr__", &CustomStr::str)
        .def_loop_cast([](const CustomStr&) -> double {
            py::pybind11_fail("Cannot cast");
        })
        .def(py::self == py::self);

    // Not explicitly convertible: `double`
    auto ss_str = [](const SimpleStruct& self) {
        return py::str("SimpleStruct({})").format(self.value);
    };
    py::dtype_user<SimpleStruct> simple_struct(m, "SimpleStruct");
    py::dtype_user<Custom> custom(m, "Custom");
    simple_struct
        .def(py::init<double>())
        .def("__str__", ss_str)
        .def("__repr__", ss_str)
        // Test operator ordering.
        .def_loop(py::self + py::self)
        .def_loop(double{} + py::self)
        .def_loop(py::self + double{})
        .def_loop_cast([](const SimpleStruct& in) -> Custom { return in; }, true);

    custom
        .def(py::init())
        .def(py::init<double>())
        .def(py::init<double, string>())
        .def(py::init<const SimpleStruct&>())
        .def(py::init<Custom>())
        .def("__repr__", [](const Custom* self) {
            return py::str("Custom({}, '{}')").format(self->value(), self->str());
        })
        .def("__str__", [](const Custom* self) {
            return py::str("C<{}, '{}'>").format(self->value(), self->str());
        })
        .def("value", &Custom::value)
        // Test referencing.
        .def("self", [](Custom* self) { return self; }, py::return_value_policy::reference)
        // Casting.
        // N.B. For `np.ones`, we could register a converter from `int64_t` to `Custom`, but this would cause a segfault,
        // because `np.ones` uses `np.copyto(..., casting="unsafe")`, which does *not* respect NPY_NEEDS_INITIALIZATION.
        // - Explicit casting (e.g., we have additional arguments).
        .def_loop_cast([](const Custom& in) { return in.value(); })
        .def_loop_cast([](const double& in) -> Custom { return in; })
            // - Implicit coercion + conversion
        .def_loop_cast(&Custom::operator SimpleStruct, true)
           // - - N.B. This shouldn't be a normal operation (upcasting?), as it may result in data loss.
        .def_loop_cast([](const SimpleStruct& in) -> Custom { return in; }, true)
            // TODO(eric.cousineau): Figure out type for implicit coercion.
            // Operators + ufuncs, with some just-operators (e.g. in-place)
        .def_loop(py::self + py::self)
        .def(py::self += py::self)
        .def_loop(py::self + double{})
        .def_loop(double{} + py::self)
        .def_loop(py::self / py::self)
        .def(py::self += double{})
        .def_loop(py::self * py::self)
        .def_loop(py::self - py::self)
        .def_loop(-py::self)
        .def_loop(py::self == py::self)
        .def_loop(py::self < py::self)
        .def(py::self + ObjectA{})
        .def(py::self + ObjectB{})
        .def_loop(py::dtype_method::dot())
        .def_loop("__pow__", [](const Custom& a, const Custom& b) {
            return CustomStr("%g ^ %g", a.value(), b.value());
        })
        // TOOD(eric.cousineau): Handle pointers too.
        .def_loop("cos", [](const Custom& self) {
            return Custom(cos(self.value()));
        })
        .def_loop("logical_and", [](const Custom&, const Custom&) -> double {
            return 10;
        });
    // Somewhat more expressive.

    auto binary_op = [](const Custom& a, const Custom& b) {
        return CustomStr("%g == %g", a.value(), b.value());
    };
    m.def("binary_op", binary_op);
    py::ufunc(m, "binary_op_loop")
        .def_loop<Custom>(binary_op);

    // N.B. We should not define a boolean operator for `equal`, as NumPy will
    // use this, even if we define it "afterwards", due to how it is stored.

    // `py::vectorize` does not seem to allow custom types due to sfinae constraints :(
    py::ufunc x(m, "same");
    x
        .def_loop<Custom>([](const Custom& a, const Custom& b) {
            return a.same_as(b);
        })
        .def_loop<CustomStr>([](const CustomStr& a, const CustomStr& b) {
            return a == b;
        })
        // Define this for checking other stuff.
        .def_loop<bool>([](bool a, bool b) { return a == b; })
        .def_loop<double>([](double a, double b) { return a == b; })
        .def_loop<py::object>([m](py::object a, py::object b) {
            // Will recurse if we don't have an overload for the given type.
            return m.attr("same")(a, b).cast<bool>();
        });
}

#else  // defined(PYBIND11_CPP14)

TEST_SUBMODULE(numpy_dtype_user, m) {
    m.attr("DISABLED") = true;
}

#endif  // defined(PYBIND11_CPP14)

File: tests/test_numpy_dtypes_user.py
import pytest

from pybind11_tests import ConstructorStats
from pybind11_tests import numpy_dtype_user as m

np = None
with pytest.suppress(ImportError):
    import numpy as np
    import sys
    sys.stderr.write("numpy version: {} {}\n".format(
        np.version.full_version, np.version.git_revision))

pytestmark = pytest.mark.skipif(
    not np or hasattr(m, "DISABLED"), reason="requires numpy and C++ >= 14")

def test_scalar_meta():
    """Tests basic metadata."""
    assert issubclass(m.Custom, np.generic)
    assert isinstance(np.dtype(m.Custom), np.dtype)


def test_scalar_ctor():
    """Tests instance lifetime management since we had to redo the instance
    registry to inherit from `np.generic` :( """
    c = m.Custom()
    c1 = m.Custom(c)
    assert id(c) != id(c1)
    assert id(c.self()) == id(c)
    del c
    del c1
    pytest.gc_collect()
    stats = ConstructorStats.get(m.Custom)
    assert stats.alive() == 0


def test_scalar_op():
    """Tests scalar operators."""
    a = m.Custom(1)
    assert repr(a) == "Custom(1.0, '')"
    assert str(a) == "C<1.0, ''>"
    assert m.same(a, a)
    b = m.Custom(2)
    assert not m.same(a, b)
    # Implicit casting is not easily testable here; see array tests.
    # Operators.
    # - self + self
    assert m.same(a + b, m.Custom(3))
    a += b
    assert m.same(a, m.Custom(3))
    # - self + double (and int, implicitly)
    assert m.same(a + 2, m.Custom(5))
    print(2 + a)
    assert m.same(2 + a, m.Custom(5))
    a += 2.
    a += 1
    assert m.same(a, m.Custom(6))
    a = m.Custom(6)
    # Others.
    assert m.same(a * b, m.Custom(12))
    assert m.same(a - b, m.Custom(4))
    assert m.same(-a, m.Custom(-6))
    # Logical.
    assert m.same(a == b, m.CustomStr("6 == 2 && '' == ''"))
    assert m.same(a < b, False)


def test_array_creation():
    # Zeros.
    x = np.zeros((2, 2), dtype=m.Custom)
    assert x.shape == (2, 2)
    assert x.dtype == m.Custom
    # Generic creation.
    x = np.array([m.Custom(1, "Howdy")])
    assert x.dtype == m.Custom
    # - Limitation on downcasting when mixing types.
    # This could be alleviated by allowing doubles to be implicitly casted for
    # the type, but it's best to avoid that.
    x = np.array([m.Custom(10), 1.])
    assert x.dtype == object
    # - At present, we will be leaking memory. This doesn't show up in instance
    # count, since these items are only mutated via `operator=`; however, it will
    # still be the case for resizing.
    # See https://github.com/numpy/numpy/issues/10721 for more information.


def test_array_creation_extended():
    with pytest.raises(ValueError):
        # Fails due to shenanigans with `np.copyto`.
        x = np.ones((2, 2), dtype=m.Custom)
    x = np.ones((1, 2)).astype(m.Custom)
    assert check_array(x, [[m.Custom(1), m.Custom(1)]])
    x = np.full((1, 2), m.Custom(10), dtype=m.Custom)
    assert check_array(x, [[m.Custom(10), m.Custom(10)]])
    # `np.eye(..., dtype=m.Custom)` requires a converter from `int` to `m.Custom` (or something that
    # tells it to use `double`). Prefer to avoid, as it complicates other implicit conversions,
    # which in general shouldn't be there, but nonetheless should be tested.
    x = np.eye(2).astype(m.Custom)
    assert check_array(x, [[m.Custom(1), m.Custom(0)], [m.Custom(0), m.Custom(1)]])


def check_array(actual, expected):
    """Checks if two arrays are exactly similar (shape, type, and data)."""
    expected = np.array(expected)
    if actual.shape != expected.shape:
        return False
    if not m.same(actual, expected).all():
        return False
    if actual.dtype != expected.dtype:
        return False
    return True


def test_array_cast():
    def check(x, dtype):
        dx = x.astype(dtype)
        assert dx.dtype == dtype, dtype
        assert dx.astype(m.Custom).dtype == m.Custom
        return dx
    # Custom -> {Custom, float, object}
    x = np.array([m.Custom(1)])
    check(x, m.Custom)
    check(x, float)
    check(x, object)
    # float -> Custom
    x = np.array([1., 2.])
    check(x, m.Custom)
    # object -> Custom
    # - See notes in the C++ code for defining the ufunc cast for `object` to
    # `Class`.
    x = np.array([1., m.Custom(10), m.SimpleStruct(100)], dtype=object)
    dx = check(x, m.Custom)
    assert check_array(dx, [m.Custom(1), m.Custom(10), m.Custom(100)])


def test_array_cast_implicit():
    a = np.array([1.]).astype(m.Custom)
    # - We registered `Custom{} + double{}`.
    a += 2
    assert check_array(a, [m.Custom(3.)])
    # - Try multiple elements.
    a = np.array([1., 2.]).astype(m.Custom)
    a += 2.
    assert check_array(a, [m.Custom(3.), m.Custom(4.)])
    # We do not allow implicit coercion for `double`:
    with pytest.raises(TypeError):
        a[0] = 1.
    with pytest.raises(TypeError):
        b = np.array([1., 2.], dtype=m.Custom)
    with pytest.raises(TypeError):
        a *= 2
    # Try an implicit conversion.
    a = np.array([1.]).astype(m.Custom)
    # - Nominal pybind implicit conversion
    a[0] = m.SimpleStruct(9)
    assert check_array(a, [m.Custom(9)])
    # - Test array construction (numpy coercion)
    c = np.array([m.SimpleStruct(10), m.SimpleStruct(11)], dtype=m.Custom)
    assert check_array(c, [m.Custom(10), m.Custom(11)])
    # - Test implicit cast via coercion.
    c *= m.SimpleStruct(2)
    assert check_array(c, [m.Custom(20), m.Custom(22)])
    # - Show dangers of implicit conversion (info loss).
    ds = m.Custom(100, "Hello")
    d = np.array([ds])
    e = np.array([m.SimpleStruct(0)])
    e[:] = d
    d[:] = e
    assert not check_array(d, [ds])
    assert check_array(d, [m.Custom(100)])


def test_array_ufunc():
    x = np.array([m.Custom(4)])
    y = np.array([m.Custom(2, "World")])
    assert check_array(x + y, [m.Custom(6)])
    # assert check_array(x + 1, [m.Custom(5)])
    assert check_array(x * y, [m.Custom(8)])
    assert check_array(x - y, [m.Custom(2)])
    assert check_array(-x, [m.Custom(-4)])
    assert check_array(x == y, [m.CustomStr("4 == 2 && '' == 'World'")])
    assert check_array(x < y, [False])
    assert check_array(np.power(x, y), [m.CustomStr("4 ^ 2")])
    assert check_array(np.dot(x, y), m.Custom(8))
    assert check_array(np.dot([x], [y]), [[m.Custom(8)]])
    print(x[0].cos())
    print(np.cos(x))


sys.stdout = sys.stderr
